include/path_tracking_pid/details/no_copy_no_move.hpp

```diff
@@ -0,0 +1,20 @@
+#pragma once
+
+namespace path_tracking_pid::details
+{
+
+// Helper class to mark a class as non-copyable and non-moveable. Use private inheritance.
+class NoCopyNoMove
+{
+public:
+  NoCopyNoMove(const NoCopyNoMove&) = delete;
+  NoCopyNoMove(NoCopyNoMove&&) = delete;
+  NoCopyNoMove& operator=(const NoCopyNoMove&) = delete;
+  NoCopyNoMove& operator=(NoCopyNoMove&&) = delete;
+
+protected:
+  NoCopyNoMove() = default;
+  ~NoCopyNoMove() = default;
```

Comment:
Why does a `NoCopyNoMove` class need a default constructor?

Reply:
It needs a default constructor because it is used as a base class, so it needs to be instantiated (without any arguments).

We need to explicitly default it because we declared (deleted) the copy and move constructors which inhibits the compiler from generating a default constructor itself.

Reply:
The compiler will generate a default constructor but only under certain conditions. And because we declared the copy c'tor and move c'tor (with ` = delete;`), it is not allowed to do so in this case.

I tried to explain that with:

> We need to explicitly default it because we declared (deleted) the copy and move constructors which inhibits the compiler from generating a default constructor itself.

Reply:
In my reply I talked about the *destructor*

Reply:
Oopsie. Indeed the destructor is always available. But we want it to be protected, so we have to default it explicitly.

Base classes should always have a (public) virtual destructor or a protected non-virtual destructor.

Reply:
OK, clear!